File: backend/src/main/java/com/taptaza/service/UserService.java
package com.taptaza.service;

import com.taptaza.dto.request.CreateAddressRequest;
import com.taptaza.dto.request.UpdateAddressRequest;
import com.taptaza.dto.request.UpdateUserRequest;
import com.taptaza.dto.response.AddressResponse;
import com.taptaza.dto.response.UserResponse;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for user and address management operations.
 */
public interface UserService {

    /**
     * Get user by ID.
     *
     * @param userId the user ID
     * @return the user response DTO
     */
    UserResponse getCurrentUser(UUID userId);

    /**
     * Update user profile (firstName, lastName).
     *
     * @param userId the user ID
     * @param request the update request containing new values
     * @return the updated user response DTO
     */
    UserResponse updateUser(UUID userId, UpdateUserRequest request);

    /**
     * Get all addresses for a user.
     *
     * @param userId the user ID
     * @return list of address response DTOs
     */
    List<AddressResponse> getAddresses(UUID userId);

    /**
     * Create a new address for a user.
     *
     * @param userId the user ID
     * @param request the create address request
     * @return the created address response DTO
     */
    AddressResponse createAddress(UUID userId, CreateAddressRequest request);

    /**
     * Update an existing address for a user.
     *
     * @param userId the user ID
     * @param addressId the address ID
     * @param request the update address request
     * @return the updated address response DTO
     */
    AddressResponse updateAddress(UUID userId, UUID addressId, UpdateAddressRequest request);

    /**
     * Delete an address for a user.
     *
     * @param userId the user ID
     * @param addressId the address ID
     */
    void deleteAddress(UUID userId, UUID addressId);
}

File: frontend/src/services/api/userService.ts
import apiClient from './client';
import {
  UserResponse,
  UpdateUserRequest,
  AddressResponse,
  CreateAddressRequest,
  UpdateAddressRequest,
} from './types';

class UserService {
  /**
   * Get current user's profile
   */
  async getProfile(): Promise<UserResponse> {
    const response = await apiClient.get<UserResponse>('/users/me');
    return response.data;
  }

  /**
   * Update current user's profile
   */
  async updateProfile(data: UpdateUserRequest): Promise<UserResponse> {
    const response = await apiClient.put<UserResponse>('/users/me', data);
    return response.data;
  }

  // ==================== ADDRESSES ====================

  /**
   * Get user's saved addresses
   */
  async getAddresses(): Promise<AddressResponse[]> {
    const response = await apiClient.get<AddressResponse[]>('/users/me/addresses');
    return response.data;
  }

  /**
   * Create a new address
   */
  async createAddress(data: CreateAddressRequest): Promise<AddressResponse> {
    const response = await apiClient.post<AddressResponse>('/users/me/addresses', data);
    return response.data;
  }

  /**
   * Update an address
   */
  async updateAddress(id: string, data: UpdateAddressRequest): Promise<AddressResponse> {
    const response = await apiClient.put<AddressResponse>(`/users/me/addresses/${id}`, data);
    return response.data;
  }

  /**
   * Delete an address
   */
  async deleteAddress(id: string): Promise<void> {
    await apiClient.delete(`/users/me/addresses/${id}`);
  }

  /**
   * Set address as default
   */
  async setDefaultAddress(id: string): Promise<AddressResponse> {
    const response = await apiClient.put<AddressResponse>(`/users/me/addresses/${id}/default`);
    return response.data;
  }
}

export const userService = new UserService();
export default userService;
